### ISSUE TITLE
Last host of hostfile not parsed when there is no EOL character

### ISSUE THREAD

**Issue by Ezibenroc** [simgrid/simgrid#136]
When there is no end-of-line character at the end of the hostfile, then the last host seems to not be parsed. No error message is displayed.

As a consequence, if the hostfile has N hosts (but the last one misses an EOL) and we launch an application with N processes, it causes two processes to go into a same host, since simgrid only parsed N-1 hosts. We thus have unexpected computation and communication times, since two of the processes share their resources.

I think it could save some debugging if simgrid at least display a warning message when the hostfile is not properly formated (I have just spent a few hours to understand what was going wrong).

-----------

Here is a full example to reproduce this. I use simgrid version `3.13.90-prealpha`.

File `test.c`, we compute fibonacci(n) independently on every process, where n is given as argument:
```c
#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>
#include <stdint.h>

uint64_t fib(int n) {
    if(n < 2)
        return 1;
    else
        return fib(n-1) + fib(n-2);
}

void compute(int rank, int size) {
    double tmp_time, computation_time;
    tmp_time = MPI_Wtime();
    uint64_t result = fib(size);
    computation_time = MPI_Wtime() - tmp_time;
    printf("rank: %4d | computation_time: %.8lf | result: %lu\n", rank, computation_time, result);
}


int main(int argc, char *argv[]) {
    MPI_Init(&argc, &argv);
    int size = 0;

    if (argc < 2) {
        fprintf(stderr, "Missing <size> argument\n"); 
        exit(1);
    } else {
        size = atoi(argv[1]);
    }

    int rank;
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    compute(rank, size);

    MPI_Finalize();
    return 0;
}
```

File `cluster_4.xml`:
```xml
<?xml version='1.0'?>
<!DOCTYPE platform SYSTEM "http://simgrid.gforge.inria.fr/simgrid/simgrid.dtd">
<platform version="4">
    <AS id="AS0" routing="Full">
        <cluster id="cluster" prefix="host-" suffix=".hawaii.edu" radical="0-3" speed="200Gf" bw="100Gbps" lat="20us"/>
    </AS>
</platform>
```

File `hostfile_4.txt`:
```txt
host-0.hawaii.edu
host-1.hawaii.edu
host-2.hawaii.edu
host-3.hawaii.edu
```

To get a faultive hostfile:
```bash
head -c -1 hostfile_4.txt > hostfile_4_faultive.txt
```

Then, compile with:
```bash
smpicc -O4 test.c -o test
```

Run with:
```bash
smpirun --cfg=smpi/running-power:6217956542.969 -np 4 -hostfile ./hostfile_4.txt -platform ./cluster_4.xml ./test 45
```

We get the “normal” behavior:
```
rank:    3 | computation_time: 0.08542052 | result: 1836311903
rank:    2 | computation_time: 0.08579984 | result: 1836311903
rank:    0 | computation_time: 0.08580138 | result: 1836311903
rank:    1 | computation_time: 0.08607242 | result: 1836311903
```

Replace the hostfile by the faultive one to see different results. Then we get this:
```
rank:    1 | computation_time: 0.08767023 | result: 1836311903
rank:    2 | computation_time: 0.08768582 | result: 1836311903
rank:    3 | computation_time: 0.17019559 | result: 1836311903
rank:    0 | computation_time: 0.17134834 | result: 1836311903
```

**Comment by adegomme** [comment]
Hello,

Thanks for your report.

Indeed I can reproduce with my latest version. Giving -map to smpirun actually shows the process/node mapping and is quite useful in this case

with hostfile_4.txt
[rank 0] -> host-0.hawaii.edu
[rank 1] -> host-1.hawaii.edu
[rank 2] -> host-2.hawaii.edu
[rank 3] -> host-3.hawaii.edu

with hostfile_4_fautive.txt
[rank 0] -> host-0.hawaii.edu
[rank 1] -> host-1.hawaii.edu
[rank 2] -> host-2.hawaii.edu
[rank 3] -> host-0.hawaii.edu

I will try to find the cause, now.

**Comment by adegomme** [comment]
So, I found the mistake. We actually went around the issue of the trailing \n at the end of the file once in smpirun, with even a comment saying "don't use wc-l for this".... And then we ignored it later on, and used wc -l anyway.

So I pushed a fix, which should be synchronized with github very soon.

**Comment by Ezibenroc** [comment]
Thank you for your quick answer!

**Comment by mquinson** [comment]
Fixed in 22ad42bc0e39c33058fa8f47d3b41966a45ffa80, thus closing.
